mechanicalsoup/stateful_browser.py

```diff
@@ -237,7 +237,7 @@ def find_link(self, *args, **kwargs):
         else:
             return links[0]
 
-    def follow_link(self, link=None, *args, **kwargs):
+    def follow_link(self, link=None, return_url_only=False, *args, **kwargs):
```

Comment:
I mentioned it in the other PR, and I still think it would be confusing for `follow_link` to have an option that makes it not follow a link and instead do something that seems more like a behavior of `find_link`.

I would be strongly in favor of removing this and changing the implementation of `download_link` (see my suggestion below), unless you can convince me that this new option is a good idea!

Reply:
I agreed with your point about not having a `download_to` parameter, because this would be a feature that the end user would really want to use and hiding it within the list of parameters of `follow_link` made it too hard to discover.

I disagree with this case: it's just a "dry-run" parameter, that asks the function to evaluate its functional part but to avoid doing side effects. I preferred calling it `return_url_only` to `dry_run` because the later doesn't say what the function is returning.

It's essentially a private feature, the user doesn't need to know about it (I still documented it in the public API in case someone wanted to use it, but we could also strip the doc and call it `_return_url_only`).

Reply:
I agreed with your point about not having the download functionality hidden in the list of arguments of `follow_link`, but this one is different. It's not an option really meant for the end user (I still documented it just in case, but we could as well strip the doc and call it `_return_url_only` to make it clear that it's not meant for the end user).

It's basically a "dry run" option: it tells the function do evaluate its functional part but to avoid doing any side effects. I preferred calling it `return_url_only` to `dry_run` because the later doesn't tell what the function would return.

Reply:
I have two primary concerns about such a parameter.

First, one might ask why we have a `dry_run`-like parameter for `follow_link`, but nothing else? Along the same lines, if I saw a `verbose` or `debug` argument for one function, I might naturally expect it to exist for most other functions, because it represents an infrastructural design choice or philosophy. But in this case, it is just one ad-hoc parameter.

Second, let us consider the possible scenarios for such a parameter:
1. The input `link` is a bs4 tag, in which case `follow_link` becomes an opaque wrapper for `link['href']`. 
2. The input `link` is not a bs4 tag, in which case `follow_link` becomes a wrapper for `find_link` that returns `link['href']` instead of `link`.

Neither of these scenarios seem like appropriate behavior for a method called `follow_link`, especially when the same functionality can be achieved using existing methods. I don't like the prospect of distorting the public API to fulfill an internal need. But even if the parameter were private, it just doesn't seem a logical thing to do.

I don't know, am I being too difficult about something that ultimately doesn't matter?

Reply:
Actually, there's another rather simple option: move most of the logic of `follow_link` (launching a browser in debug mode and special-casing elements with an href attribute) to `find_link`. Then `find_link` becomes a straightforward wrapper, and `download_link` remains simple too. I'll push something along these lines shortly.